hw/drivers/flash/spiflash/src/spiflash.c

```diff
@@ -0,0 +1,349 @@
+        uint32_t *address, uint32_t *sz);
+
+static const struct hal_flash_funcs spiflash_flash_funcs = {
+    .hff_read         = spiflash_read,
+    .hff_write        = spiflash_write,
+    .hff_erase_sector = spiflash_erase_sector,
+    .hff_sector_info  = spiflash_sector_info,
+    .hff_init         = spiflash_init,
+};
+
+struct spiflash_dev spiflash_dev = {
+    /* struct hal_flash for compatibility */
+    .hal = {
+        .hf_itf        = &spiflash_flash_funcs,
+        .hf_base_addr  = 0,
+        .hf_size       = MYNEWT_VAL(SPIFLASH_SECTOR_COUNT) *
+                         MYNEWT_VAL(SPIFLASH_SECTOR_SIZE),
+        .hf_sector_cnt = MYNEWT_VAL(SPIFLASH_SECTOR_COUNT),
+        .hf_align      = 0,
+    },
+
+    /* SPI settings */
+    .spi_settings = {
+        .data_order = HAL_SPI_MSB_FIRST,
+        .data_mode  = HAL_SPI_MODE3,
+        .baudrate   = MYNEWT_VAL(SPIFLASH_BAUDRATE),
+        .word_size  = HAL_SPI_WORD_SIZE_8BIT,
+    },
+
+    .sector_size        = MYNEWT_VAL(SPIFLASH_SECTOR_SIZE),
+    .page_size          = MYNEWT_VAL(SPIFLASH_PAGE_SIZE),
+    .spi_num            = MYNEWT_VAL(SPIFLASH_SPI_NUM),
+    .spi_cfg            = NULL,
+    .ss_pin             = MYNEWT_VAL(SPIFLASH_SPI_CS_PIN),
+};
+
+static inline void
+spiflash_cs_activate(struct spiflash_dev *dev)
+{
+    hal_gpio_write(dev->ss_pin, 0);
+}
+
+static inline void
+spiflash_cs_deactivate(struct spiflash_dev *dev)
+{
+    hal_gpio_write(dev->ss_pin, 1);
+}
+
+uint8_t
+spiflash_release_power_down(struct spiflash_dev *dev, uint8_t *id)
+{
+    uint8_t cmd[5] = { SPIFLASH_RELEASE_POWER_DOWN, 0xFF, 0xFF, 0xFF, 0 };
+
+    spiflash_cs_activate(dev);
+
+    hal_spi_txrx(dev->spi_num, cmd, cmd, sizeof cmd);
+
+    spiflash_cs_deactivate(dev);
+
+    if (id)
+        *id = cmd[4];
+
+    return 0;
+}
+
+uint8_t
+spiflash_read_jedec_id(struct spiflash_dev *dev,
+        uint8_t *manufacturer, uint8_t *memory_type, uint8_t *capacity)
+{
+    uint8_t cmd[4] = { SPIFLASH_READ_JEDEC_ID, 0, 0, 0 };
+
+    spiflash_cs_activate(dev);
+
+    hal_spi_txrx(dev->spi_num, cmd, cmd, sizeof cmd);
+
+    spiflash_cs_deactivate(dev);
+
+    if (manufacturer)
+        *manufacturer = cmd[1];
+
+    if (memory_type)
+        *memory_type = cmd[2];
+
+    if (capacity)
+        *capacity = cmd[3];
+
+    return 0;
+}
+
+uint8_t
+spiflash_read_status(struct spiflash_dev *dev)
+{
+    uint8_t val;
+
+    spiflash_cs_activate(dev);
+
+    hal_spi_tx_val(dev->spi_num, SPIFLASH_READ_STATUS_REGISTER);
+    val = hal_spi_tx_val(dev->spi_num, 0xFF);
+
+    spiflash_cs_deactivate(dev);
+
+    return val;
+}
+
+bool
+spiflash_device_ready(struct spiflash_dev *dev)
+{
+    return !(spiflash_read_status(dev) & SPIFLASH_STATUS_BUSY);
+}
+
+int
+spiflash_wait_ready(struct spiflash_dev *dev, uint32_t timeout)
+{
+    os_time_t exp_time = os_time_get() + timeout;
+    while (!spiflash_device_ready(dev)) {
+        if (os_time_get() > exp_time)
```

Comment:
What this means is that the wait-for-completion timeout depends on current OS tick rate. Which is cool, but you should then make the parameter to this call depend on that. OS_TICKS_PER_SEC is the define to compute it with.